graph/api/src/main/scala/org/apache/spark/graph/api/GraphElementFrame.scala

```diff
@@ -0,0 +1,260 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.graph.api
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.sql.DataFrame
+
+/**
+ * A [[PropertyGraph]] is created from GraphElementFrames.
+ *
+ * A graph element is either a node or a relationship.
+ * A GraphElementFrame wraps a DataFrame and describes how it maps to graph elements.
+ *
+ * @since 3.0.0
+ */
+abstract class GraphElementFrame {
+
+  /**
+   * Initial DataFrame that can still contain unmapped, arbitrarily ordered columns.
+   *
+   * @since 3.0.0
+   */
+  def df: DataFrame
+
+  /**
+   * Name of the column that contains the graph element identifier.
+   *
+   * @since 3.0.0
+   */
+  def idColumn: String
+
+  /**
+   * Name of all columns that contain graph element identifiers.
+   *
+   * @since 3.0.0
+   */
+  def idColumns: Seq[String] = Seq(idColumn)
+
+  /**
+   * Mapping from graph element property keys to the columns that contain the corresponding property
+   * values.
+   *
+   * @since 3.0.0
+   */
+  def properties: Map[String, String]
+
+}
+
+object NodeFrame {
+
+  /**
+   * Describes how to map an initial DataFrame to nodes.
+   *
+   * All columns apart from the given `idColumn` are mapped to node properties.
+   *
+   * @param df        DataFrame containing a single node in each row
+   * @param idColumn  column that contains the node identifier
+   * @param labelSet  labels that are assigned to all nodes
+   * @since 3.0.0
+   */
+  def create(df: DataFrame, idColumn: String, labelSet: Set[String]): NodeFrame = {
+    val properties = (df.columns.toSet - idColumn)
+      .map(columnName => columnName -> columnName)
+      .toMap
+    create(df, idColumn, labelSet, properties)
+  }
+
+  /**
+   * Describes how to map an initial DataFrame to nodes.
+   *
+   * All columns apart from the given `idColumn` are mapped to node properties.
+   *
+   * @param df        DataFrame containing a single node in each row
+   * @param idColumn  column that contains the node identifier
+   * @param labelSet  labels that are assigned to all nodes
+   * @param properties mapping from property keys to corresponding columns
+   * @since 3.0.0
+   */
+  def create(
+      df: DataFrame,
+      idColumn: String,
+      labelSet: Set[String],
+      properties: Map[String, String]): NodeFrame = {
```

Comment:
* It is hard to guess from the `properties` name that the map values are column names.
* Users can use `select` and alias to drop/rename columns before creating this DataFrame. So do we still need this properties field?

Reply:
Mh, I think having convenience methods for that is actually very helpful for the user. Imho, the documentation makes the purpose of that Map clear, we could however rename the field to `propertiesToColumns` or `propertyColumns`? (I prefer the latter).